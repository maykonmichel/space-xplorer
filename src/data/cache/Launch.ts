import {FieldPolicy, TypePolicy} from '@apollo/client';
import {format} from 'date-fns';

const launch_date_formatted: FieldPolicy<string> = {
  read: (existing, {readField}) => {
    const date = readField<string>('launch_date_local') || '';
    return format(new Date(date), 'MM/dd/yyyy hh:mmaaa');
  },
};

const Launch: TypePolicy = {
  fields: {
    launch_date_formatted,
  },
};

export default Launch;
