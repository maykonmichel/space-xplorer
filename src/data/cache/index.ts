import {InMemoryCache} from '@apollo/client';
import {format} from 'date-fns';

import {initializeFavoriteImages} from '~/data/favoriteImages';

const cache = new InMemoryCache({
  typePolicies: {
    Launch: {
      fields: {
        launch_date_formatted: {
          read: (existing, {readField}) => {
            const unix = readField<number>('launch_date_unix') || 0;
            return format(unix, 'MM/dd/yyyy hh:mmaaa');
          },
        },
      },
    },
    LaunchLinks: {
      merge: true,
    },
  },
});

export const initializeCache = () => {
  initializeFavoriteImages(cache);
};

initializeCache();

export default cache;
