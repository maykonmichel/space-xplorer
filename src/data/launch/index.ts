import {gql, useQuery} from '@apollo/client';

type LaunchLinks = {
  article_link: string;
  flickr_images: string[];
};

type LaunchRocket = {
  rocket_name: string;
};

export type Launch = {
  id: string;
  details: string;
  launch_date_local: string;
  launch_date_formatted: string;
  links: LaunchLinks;
  mission_name: string;
  rocket: LaunchRocket;
};

const QUERY = gql`
  query ($id: ID!) {
    launch(id: $id) {
      id
      details
      launch_date_formatted @client
      launch_date_local
      links {
        article_link
        flickr_images
      }
      mission_name
      rocket {
        rocket_name
      }
    }
  }
`;

type Data = {
  launch: Launch;
};

type Variables = {
  id: string;
};

export const useLaunchQuery = (id: string) =>
  useQuery<Data, Variables>(QUERY, {variables: {id}});
