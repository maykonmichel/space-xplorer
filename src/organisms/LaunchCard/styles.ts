import {StyleSheet} from 'react-native';

import colors from '~/theme/colors';

export default StyleSheet.create({
  button: {
    margin: 16,
    color: colors.secondary.light,
    textTransform: 'uppercase',
    fontWeight: 'bold',
  },
  container: {
    marginHorizontal: 16,
    marginVertical: 8,
    borderRadius: 4,
    backgroundColor: colors.neutral.light,
  },
  date: {
    marginHorizontal: 16,
    color: colors.neutral.medium,
  },
  image: {
    aspectRatio: 1920 / 1080,
    width: '100%',
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
  },
  title: {
    margin: 16,
    marginBottom: 4,
    color: colors.neutral.dark,
    fontSize: 24,
  },
});
