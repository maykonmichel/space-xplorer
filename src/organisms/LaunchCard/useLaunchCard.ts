import {useMainStackNavigation} from '~/navigators/MainStack';
import {LAUNCH} from '~/screens';

export default (id: string) => {
  const {navigate} = useMainStackNavigation();

  const onPress = () => navigate(LAUNCH, {id});

  return {
    onPress,
  };
};
