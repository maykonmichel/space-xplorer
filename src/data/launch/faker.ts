import faker from 'faker';
import {repeat} from 'ramda';

import {Launch} from '~/data/launch';

export const fakerFlickrImages = (): Launch['links']['flickr_images'] =>
  repeat(faker.image.image(), faker.datatype.number({min: 1, max: 5}));

export const fakerLaunch = (): Launch => ({
  id: faker.datatype.number().toString(),
  details: faker.lorem.paragraph(),
  mission_name: faker.lorem.words(),
  launch_date_formatted: faker.date.past().toLocaleDateString(),
  launch_date_local: faker.date.past().toLocaleDateString(),
  links: {
    article_link: faker.internet.url(),
    flickr_images: fakerFlickrImages(),
  },
  rocket: {
    rocket_name: faker.lorem.words(),
  },
});
