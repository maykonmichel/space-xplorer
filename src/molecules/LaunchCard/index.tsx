import React, {FC} from 'react';
import {Image, Text, TouchableOpacity} from 'react-native';

import {Launch} from '~/data/launch';
import {useMainStackNavigation} from '~/navigators/MainStack';
import {LAUNCH} from '~/screens';

const LaunchCard: FC<Launch> = ({
  id,
  mission_name,
  launch_date_local,
  links: {flickr_images},
}) => {
  const {navigate} = useMainStackNavigation();

  const onPress = () => navigate(LAUNCH, {id});

  const [uri] = flickr_images;

  return (
    <TouchableOpacity onPress={onPress}>
      <Text>{mission_name}</Text>
      <Text>{launch_date_local}</Text>
      {uri && <Image source={{uri}} style={{width: 200, height: 100}} />}
    </TouchableOpacity>
  );
};

export default LaunchCard;
