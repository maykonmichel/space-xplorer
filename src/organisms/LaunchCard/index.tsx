import React, {FC} from 'react';
import {Image, Text, TouchableOpacity} from 'react-native';

import {Launch} from '~/data/launch';
import {useMainStackNavigation} from '~/navigators/MainStack';
import {LAUNCH} from '~/screens';

import styles from './styles';

export type Props = Launch;

const LaunchCard: FC<Props> = ({
  id,
  mission_name,
  launch_date_local,
  links: {flickr_images},
}) => {
  const {navigate} = useMainStackNavigation();

  const onPress = () => navigate(LAUNCH, {id});

  const [uri] = flickr_images;

  return (
    <TouchableOpacity onPress={onPress}>
      <Text>{mission_name}</Text>
      <Text>{launch_date_local}</Text>
      {uri && <Image source={{uri}} style={styles.image} />}
    </TouchableOpacity>
  );
};

export default LaunchCard;
