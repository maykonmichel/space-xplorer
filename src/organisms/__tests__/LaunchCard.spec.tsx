import {MockedProvider} from '@apollo/client/testing';
import {render} from '@testing-library/react-native';
import faker from 'faker';
import React from 'react';

import cache from '~/data/cache';
import {fakerLaunch} from '~/data/launch/faker';
import LaunchCard, {Props} from '~/organisms/LaunchCard';

jest.mock('@react-navigation/native', () => {
  const actual = jest.requireActual('@react-navigation/native');
  return {
    ...actual,
    useNavigation: () => ({
      dispatch: jest.fn(),
    }),
  };
});

const getProps = (): Props => fakerLaunch();

const getSut = (props: Props) =>
  render(
    <MockedProvider cache={cache}>
      <LaunchCard {...props} />
    </MockedProvider>,
  );

describe('<LaunchCard />', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  it('should render as expected', async () => {
    const props = getProps();
    const {toJSON} = getSut({
      ...props,
      mission_name: 'mission name',
      launch_date_local: '01/01/1998',
      links: {
        ...props.links,
        flickr_images: [faker.image.imageUrl()],
      },
    });

    expect(toJSON()).toMatchSnapshot();
  });
});
