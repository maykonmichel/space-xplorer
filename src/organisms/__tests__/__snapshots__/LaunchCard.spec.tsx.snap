// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchCard /> should render as expected 1`] = `
<View
  accessibilityHint="See launch details"
  accessibilityLabel="Launch card"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FAFAFA",
      "borderRadius": 4,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "aspectRatio": 1.7777777777777777,
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
          "width": "100%",
        },
      ]
    }
  >
    <FastImageView
      accessibilityLabel="Launch image"
      resizeMode="cover"
      source={
        Object {
          "uri": "http://placeimg.com/640/480",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <Text
    accessibilityLabel="Launch name"
    style={
      Object {
        "color": "#212121",
        "fontSize": 24,
        "margin": 16,
        "marginBottom": 4,
      }
    }
  >
    mission name
  </Text>
  <Text
    accessibilityLabel="Launch date"
    style={
      Object {
        "color": "#9E9E9E",
        "marginHorizontal": 16,
      }
    }
  >
    01/01/1998 00:00am
  </Text>
  <Text
    style={
      Object {
        "color": "#FF77A9",
        "fontWeight": "bold",
        "margin": 16,
        "textTransform": "uppercase",
      }
    }
  >
    See more
  </Text>
</View>
`;
